tests/index/test_find2_index_approx.h

```diff
@@ -0,0 +1,486 @@
+    testOptimalSearchScheme(OptimalSearchSchemes<1, 2>::VALUE, HammingDistance());
+    testOptimalSearchScheme(OptimalSearchSchemes<1, 3>::VALUE, HammingDistance());
+    testOptimalSearchScheme(OptimalSearchSchemes<1, 4>::VALUE, HammingDistance());
+    testOptimalSearchScheme(OptimalSearchSchemes<2, 2>::VALUE, HammingDistance());
+    testOptimalSearchScheme(OptimalSearchSchemes<2, 3>::VALUE, HammingDistance());
+    testOptimalSearchScheme(OptimalSearchSchemes<2, 4>::VALUE, HammingDistance());
+    testOptimalSearchScheme(OptimalSearchSchemes<3, 3>::VALUE, HammingDistance());
+    testOptimalSearchScheme(OptimalSearchSchemes<3, 4>::VALUE, HammingDistance());
+    testOptimalSearchScheme(OptimalSearchSchemes<4, 4>::VALUE, HammingDistance());
+}
+
+// ----------------------------------------------------------------------------
+// Test test_find2_index_approx_edit
+// ----------------------------------------------------------------------------
+
+SEQAN_DEFINE_TEST(test_find2_index_approx_edit)
+{
+    testOptimalSearchScheme(OptimalSearchSchemes<0, 0>::VALUE, EditDistance());
+    testOptimalSearchScheme(OptimalSearchSchemes<0, 1>::VALUE, EditDistance());
+    testOptimalSearchScheme(OptimalSearchSchemes<0, 2>::VALUE, EditDistance());
+    testOptimalSearchScheme(OptimalSearchSchemes<0, 3>::VALUE, EditDistance());
+    testOptimalSearchScheme(OptimalSearchSchemes<0, 4>::VALUE, EditDistance());
+    testOptimalSearchScheme(OptimalSearchSchemes<1, 1>::VALUE, EditDistance());
+    testOptimalSearchScheme(OptimalSearchSchemes<1, 2>::VALUE, EditDistance());
+    testOptimalSearchScheme(OptimalSearchSchemes<1, 3>::VALUE, EditDistance());
+    testOptimalSearchScheme(OptimalSearchSchemes<1, 4>::VALUE, EditDistance());
+    testOptimalSearchScheme(OptimalSearchSchemes<2, 2>::VALUE, EditDistance());
+    testOptimalSearchScheme(OptimalSearchSchemes<2, 3>::VALUE, EditDistance());
+    testOptimalSearchScheme(OptimalSearchSchemes<2, 4>::VALUE, EditDistance());
+    testOptimalSearchScheme(OptimalSearchSchemes<3, 3>::VALUE, EditDistance());
+    testOptimalSearchScheme(OptimalSearchSchemes<3, 4>::VALUE, EditDistance());
+    testOptimalSearchScheme(OptimalSearchSchemes<4, 4>::VALUE, EditDistance());
+}
+
+SEQAN_DEFINE_TEST(test_find2_index_approx_small_test)
+{
+    DnaString genome("GAGAGGCCACTCGCAGGATTAAGTCAATAAGTTAATGGCGTCGGCTTCCTGGTATGTAGTACGACGCCCACAGTGACCTCATCGGTGCATTTCCTCATCGTAGGCGGAACGGTAGACACAAGGCATGATGTCAAATCGCGACTCCAATCCCAAGGTCGCAAGCCTATATAGGAACCCGCTTATGCCCTCTAATCCCGGACAGACCCCAAATATGGCATAGCTGGTTGGGGGTACCTACTAGGCACAGCCGGAAGCA");
+    DnaString pattern1("GGGGTTAT");
+    DnaString pattern2("CTAGCTAA");
+    StringSet<DnaString> patterns;
+    appendValue(patterns, pattern1);
+    appendValue(patterns, pattern2);
+
+    std::set<Pair<unsigned, unsigned> > hits, expectedHits {{14, 15}, {56, 57}, {63, 64}, {83, 84}, {154, 155}};
+
+    std::mutex mtx;
+    auto delegate = [&hits, &mtx](auto & it)
+    {
+        mtx.lock();
```

Comment:
I thought it is marked as a critical section already? In any case, please use lock_guards and not mutexes directly (exception safety!!).